### ISSUE TITLE
Conversion: highlightElement children consumption may not be correct

### ISSUE THREAD

**Issue by scofalik** [ckeditor/ckeditor5-engine#1395]
```javascript
// Consume element itself.
conversionApi.consumable.consume( data.item, evt.name );

// Consume all children nodes.
for ( const value of ModelRange.createIn( data.item ) ) {
	conversionApi.consumable.consume( value.item, evt.name );
}
````

This is how consumables are consumed when a marker on an element with custom highlighting is converted.

Basically, we consume the element itself and all of its children to prevent conversion inside that element. However, this is not correct... Take a look at this example:

```html
<paragraph>fo{z</paragraph><widget>ba}r</widget>
```

Assume `widget` is an element that has own highlighting and `{}` is the marker to convert. During conversion, the consumable object will have:

* `z` text proxy,
* `widget` element,
* and `ba` text proxy.

However, during `widget` conversion we won't consume `ba` text proxy. Instead, we will try to consume whole `bar` text proxy. The converter does not know that only `ba` part is in the consumables.

I see three solutions:

1. Fix the converter. _Somehow_. I don't know how yet because I haven't been thinking about it.
2. Change `ModelConsumable` so that if bigger text proxy is consumed, then the smaller part is consumed too. This is however more tricky than it sounds. Sometimes maybe we would want to stop conversion if only a part of text _to be consumed_? This may lead to other WTFs and maybe buggy behavior.
3. Leave it as it is. If not all children are in the marker, convert the children anyway. This is not the worst for comments or user selection (actually it is even better for user selection cause it better reflects what user has selected). However, it may not be good in some cases, where content inside a widget might look bad when highlighted.

**Comment by pjasiun** [comment]
Related: https://github.com/ckeditor/ckeditor5-engine/issues/350

**Comment by scofalik** [comment]
I don't think this is related. This is a different issue. The linked issue is about having one, same instance for "equal text proxies". A bit like singleton pattern. Resolving the linked issue will not help in any way with this issue.

**Comment by pjasiun** [comment]
For sure it's a bug in how `consume` works and it can be fixed. The question is, what is expected behavior if selection ends or starts in the middle of the nested editable. Maybe widget should consume markers on children only if all children are selected?

**Comment by scofalik** [comment]
Why is it a bug?

In consumable I have text proxy containing `ba`.

* If I pass text proxy `b` or `bar` from the same text node, it would not consume,
* If I pass text proxy `ba` -- no matter if the same instance or not -- it will consume.

Where's the bug?

The only thing might be some missing functionality (to consume text proxies partially).

**Comment by pjasiun** [comment]
Ok, you are right. If `r` is not consumable it should not be possible to consume `bar`. Maybe the widget should check what is consumable instead of simply using `range.createIn`. But, yea, it's tricky.

**Comment by scofalik** [comment]
There's no API for that as for now, but yes this is also another option.
